### PR TITLE
Tar: Only treat reparse points marked as junctions or symlinks as actual tar symlinks

### LINKED ISSUE #82949: TarWriter on Windows Server treats File Deduplication reparse point flag as symbolic links
https://github.com/dotnet/runtime/blob/9fd3ece652fed473c9f2398363da6d7bca41a5d9/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.Windows.cs#L25

As seen in the snippet above, the Tar Entry gets mistaken for a symlink if the reparse point flag is set, but it doesn't verify if the file/directory is actually a symlink/junction. Windows Server with file deduplication enabled, or even OneDrive synced files, can set the Reparse point flag, but utilizes entirely different reparse tags.

IO_REPARSE_TAG_DEDUP is used for deduplication, whereas IO_REPARSE_TAG_MOUNT_POINT & IO_REPARSE_TAG_SYMLINK are used for junctions and symlinks.

### PR DESCRIPTION
**Pull request by carlossanlop**
The current code only checks if it's a reparse point and adds the filesystem object as a symlink, which is wrong for many types of reparse points. Only those marked as junctions or symlinks should be treated as tar symlinks.

I added a test to confirm that a junction is stored as symlink. I am unsure how to create other types of reparse points so I didn't add tests for that, but since those are rarer cases, I would prefer to tackle them separately if requested.

cc @billfreist, thanks for reporting this.

Fixes https://github.com/dotnet/runtime/issues/82949

### CODE REVIEW

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.Windows.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+namespace System.Formats.Tar.Tests;
+
+public partial class TarWriter_WriteEntry_File_Tests : TarWriter_File_Base
+{
+    [Theory]
+    [InlineData(TarEntryFormat.V7)]
+    [InlineData(TarEntryFormat.Ustar)]
+    [InlineData(TarEntryFormat.Pax)]
+    [InlineData(TarEntryFormat.Gnu)]
+    public void Add_Junction_As_SymbolicLink(TarEntryFormat format)
```
Comment (danmoseley) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.Windows.cs:16]: Can you run these tests without elevation? I think it's missing a can create symlinks condition. I've added them a few times when tests failed locally.

Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.Windows.cs:16]: Starting with Windows 10 Insiders build 14972, symlinks can be created without needing to elevate the console as administrator. https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/

If we still have tests running older Windows versions, they should break in the CI.

Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.Windows.cs:16]: Junctions are ancient, they don't require elevation AFAIK. They ran without problem in my local machine. The CI should also cry if something goes wrong.

This new test does not create symlinks in the disk by the way, it adds an entry to the tar archive as a symlink. The only thing that gets created in disk is the junction.

**Review by ViktorHofer — APPROVED** [review]

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.FindFirstFileEx.cs [hunk 2]
```diff
@@ -24,6 +24,21 @@ internal static SafeFindHandle FindFirstFile(string fileName, ref WIN32_FIND_DAT
...
+            {
+                int errorCode = Marshal.GetLastPInvokeError();
+                // File not found doesn't make much sense coming from a directory.
+                if (isDirectory && errorCode == Errors.ERROR_FILE_NOT_FOUND)
+                    errorCode = Errors.ERROR_PATH_NOT_FOUND;
+                if (ignoreAccessDenied && errorCode == Errors.ERROR_ACCESS_DENIED)
+                    return;
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath);
+            }
+        }
```
Comment (Jozkee) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.FindFirstFileEx.cs:40]: nit: we already have a `GetFindData` method, would be nice to re-use this one there.
https://github.com/dotnet/runtime/blob/e8a18d691c4316eb20e22c00158eac5d98cee987/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs#L276-L289

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.Windows.cs [hunk 3]
```diff
@@ -21,18 +19,32 @@ private TarEntry ConstructEntryForWriting(string fullPath, string entryName, Fil
 
             FileAttributes attributes = File.GetAttributes(fullPath);
 
+            bool isDirectory = (attributes & FileAttributes.Directory) != 0;
+
+            Interop.Kernel32.WIN32_FIND_DATA data = default;
+            Interop.Kernel32.GetFindData(fullPath, isDirectory, ignoreAccessDenied: false, ref data);
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.Windows.cs:25]: Shouldn't this be called only when the file is a reparse point?

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.Windows.cs [hunk 4]
```diff
@@ -21,18 +19,32 @@ private TarEntry ConstructEntryForWriting(string fullPath, string entryName, Fil
...
+
             TarEntryType entryType;
             if ((attributes & FileAttributes.ReparsePoint) != 0)
             {
-                entryType = TarEntryType.SymbolicLink;
+                if (data.dwReserved0 is Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK or
+                                        Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_MOUNT_POINT)
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.Windows.cs:31]: This looks like another data point in favor of having an API for ReparseTags https://github.com/dotnet/runtime/issues/1908.

**Comment by Jozkee** [comment]
> I am unsure how to create other types of reparse points so I didn't add tests for that

You can test IO_REPARSE_TAG_APPEXECLINK by trying to archive one file with such ReparseTag:

https://github.com/dotnet/runtime/blob/b8653d9c40ceca92805a68a7180f2e0caee65df8/src/libraries/System.IO.FileSystem/tests/File/SymbolicLinks.cs#L49-L61

**Comment by carlossanlop** [comment]
> You can test IO_REPARSE_TAG_APPEXECLINK by trying to archive one file with such ReparseTag:

Thanks @Jozkee, I forgot about that.

I wrote a test that attempts to verify what happens when adding an AppExecLink. I reused the method that we have to retrieve the first AppExecLink from the WindowsApps folder. In my case, it's `accesschk.exe`, which is a binary that is installed by the Sysinternals Suite. I confirmed it's an actual AppExecLink because it's 0KB in length.

It seems that my current code would throw an IOException when FileStream attempts to create a file handle to it (see the callkstack below). Which means we can't treat AppExecLinks as regular files.

I think for now, we should just throw an IOException on the Tar code when we encounter an unsupported ReparsePoint. There might be a way to crack open an AppExecLink, retrieve some internal information it holds, and then store it in the DataStream, but we don't have an actual entry type to indicate this is an AppExecLink file, which differs from a regular file and from a symlink. So if the user generates a tar archive containing one of these, then tries to extract it, it will be extracted as something that it is not.

Thoughts?

```
      System.IO.IOException : The file cannot be accessed by the system. : 'C:\Users\calope\AppData\Local\Microsoft\WindowsApps\accesschk.exe'
        Stack Trace:
          C:\Users\calope\source\repos\runtime\src\libraries\System.Private.CoreLib\src\Microsoft\Win32\SafeHandles\SafeFileHandle.Windows.cs(114,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share
  , FileOptions options)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Private.CoreLib\src\Microsoft\Win32\SafeHandles\SafeFileHandle.Windows.cs(47,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileO
  ptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Private.CoreLib\src\System\IO\Strategies\OSFileStreamStrategy.cs(46,0): at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions option
  s, Int64 preallocationSize, Nullable`1 unixCreateMode)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Private.CoreLib\src\System\IO\Strategies\SyncWindowsFileStreamStrategy.cs(15,0): at System.IO.Strategies.SyncWindowsFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share,
  FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Private.CoreLib\src\System\IO\Strategies\FileStreamHelpers.Windows.cs(35,0): at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, Fil
  eOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Private.CoreLib\src\System\IO\Strategies\FileStreamHelpers.cs(24,0): at System.IO.Strategies.FileStreamHelpers.ChooseStrategy(FileStream fileStream, String path, FileMode mode, FileAccess access, FileShare
   share, Int32 bufferSize, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Private.CoreLib\src\System\IO\FileStream.cs(222,0): at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationS
  ize)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Private.CoreLib\src\System\IO\FileStream.cs(146,0): at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Formats.Tar\src\System\Formats\Tar\TarWriter.Windows.cs(79,0): at System.Formats.Tar.TarWriter.ConstructEntryForWriting(String fullPath, String entryName, FileOptions fileOptions)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Formats.Tar\src\System\Formats\Tar\TarWriter.cs(166,0): at System.Formats.Tar.TarWriter.ReadFileFromDiskAndWriteToArchiveStreamAsEntry(String fullPath, String entryName)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Formats.Tar\src\System\Formats\Tar\TarWriter.cs(139,0): at System.Formats.Tar.TarWriter.WriteEntry(String fileName, String entryName)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Formats.Tar\tests\TarWriter\TarWriter.WriteEntry.File.Tests.Windows.cs(72,0): at System.Formats.Tar.Tests.TarWriter_WriteEntry_File_Tests.Add_AppExecLink_As_RegularFile(TarEntryFormat format)
             at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
          C:\Users\calope\source\repos\runtime\src\coreclr\System.Private.CoreLib\src\System\Reflection\MethodBaseInvoker.CoreCLR.cs(36,0): at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
          C:\Users\calope\source\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodBaseInvoker.cs(177,0): at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
      System.Formats.Tar.Tests.TarWriter_WriteEntry_File_Tests.Add_AppExecLink_As_RegularFile(format: Ustar) [FAIL]
```

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.Windows.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            Assert.Equal(TarEntryType.SymbolicLink, entry.EntryType);
+            Assert.Null(entry.DataStream);
+
+            VerifyPlatformSpecificMetadata(junctionPath, entry);
+
+            Assert.Null(reader.GetNextEntry());
+        }
+    }
+
+    [Theory]
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.Windows.cs:55]: You need to use `[ConditionalTheory]` in order for SkipTestException to actually skip and not take down the whole test run.
```suggestion
    [ConditionalTheory]
```

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.Windows.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    public void Add_Unsupported_ReparsePoints_Throws(TarEntryFormat format)
+    {
+        string? appExecLinkPath = MountHelper.GetAppExecLinkPath();
+        if (appExecLinkPath == null)
+        {
+            throw new SkipTestException("Could not find an appexeclink in this machine.");
+        }
+
+        using MemoryStream archive = new MemoryStream();
+        using TarWriter writer = new TarWriter(archive, format, leaveOpen: true);
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.Windows.cs:69]:
```suggestion
        using TarWriter writer = new TarWriter(archive, format);
```

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntryAsync.File.Tests.Windows.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            Assert.Equal(TarEntryType.SymbolicLink, entry.EntryType);
+            Assert.Null(entry.DataStream);
+
+            VerifyPlatformSpecificMetadata(junctionPath, entry);
+
+            Assert.Null(await reader.GetNextEntryAsync());
+        }
+    }
+
+    [Theory]
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntryAsync.File.Tests.Windows.cs:55]:
```suggestion
    [ConditionalTheory]
```

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntryAsync.File.Tests.Windows.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    public async Task Add_Unsupported_ReparsePoints_Throws_Async(TarEntryFormat format)
+    {
+        string? appExecLinkPath = MountHelper.GetAppExecLinkPath();
+        if (appExecLinkPath == null)
+        {
+            throw new SkipTestException("Could not find an appexeclink in this machine.");
+        }
+
+        await using MemoryStream archive = new MemoryStream();
+        await using TarWriter writer = new TarWriter(archive, format, leaveOpen: true);
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntryAsync.File.Tests.Windows.cs:69]:
```suggestion
        await using TarWriter writer = new TarWriter(archive, format);
```

**Review by jozkee — APPROVED** [review]

**Comment by ericstj** [comment]
@carlossanlop what's the next step to move this one forward?

**Comment by carlossanlop** [comment]
> @carlossanlop what's the next step to move this one forward?

Make sure we have tests that confirm my changes are working, haven't been able to get back to it yet. Need to find a way to test adding a OneDrive cloud file. Maybe I could add manual tests.

**Pull request closed without merging**